torch/_torch_docs.py

```diff
@@ -13897,3 +13897,82 @@ def merge_dicts(*dicts):
 are freshly created instead of aliasing the input.
 """,
 )
+
+
+for unary_base_func_name in (
+    "exp",
+    "sqrt",
+    "abs",
+    "acos",
+    "asin",
+    "atan",
+    "ceil",
+    "cos",
+    "cosh",
+    "erf",
+    "erfc",
+    "expm1",
+    "floor",
+    "log",
+    "log10",
+    "log1p",
+    "log2",
+    "neg",
+    "tan",
+    "tanh",
+    "sin",
+    "sinh",
+    "round",
+    "lgamma",
+    "frac",
+    "reciprocal",
+    "sigmoid",
+    "trunc",
+):
+    unary_foreach_func_name = f"_foreach_{unary_base_func_name}"
+    add_docstr(
+        getattr(torch, unary_foreach_func_name),
+        r"""
+{}(self: List[Tensor]) -> List[Tensor]
+
+Apply :func:`torch.{}` to each Tensor of the input list.
+
+If the input tensors meet a certain condition, for example, they have the same dtype,
+are hosted on the same cuda device, and contiguous (including channels_last and channels_last_3d),
+then the operation done with fewer cuda kernel calls.
+
+.. note::
+    Only CPU and CUDA are supported. Forward-mode AD is not supported.
+
+Args:
+    self (list of Tensors): Input list of Tensors. Each Tensor can have an arbitrary shape, dtype, device, and strides.
```

Comment:
It works for difference dtype and device? Why do we collect grads to zero out by device/dtype here then: https://github.com/pytorch/pytorch/blob/274d3b24c3f4b6dc8964ee6642e909ce4b992112/torch/optim/optimizer.py#L394 ?

Reply:
To reduce the number of CUDA kernels launched as possible, I guess.

Reply:
I'm not sure to follow? How does the two have different number of kernel launched?

Reply:
If we pass a list of tensors of different dtypes and/or devices to a foreach function, it'll call the corresponding aten native function the number of Tensors times, i.e. len(self) cuda kernels. With the grouping, it could reduce the number to `len(per_device_and_dtype_grads)`, assuming all the tensors are a CUDA tensor

Reply:
@alband _for_each ops support args on the different device and of the different dtypes, but they'll fall back to slow implementation. If, instead of just calling `for_each` we pre-sort the args, we can call `for_each` with a subset of tensors that'll go to the fast path. Really this sorting should be done in for_each itself though, and not in the optimizer

Reply:
Ok!
Should we really advertise this then as it will not do what you expect these functions to do.